src/server/index.js

```diff
@@ -1,188 +1,228 @@
     onEnd()
   }
 }
 
+/**
+* Stardust Transport Server
+* @class
+*/
 class Server {
-  constructor ({ transports, addresses, muxers }) {
-    this.switch = new MicroSwitch({ transports, addresses, muxers, handler: this.handler.bind(this) })
-
+  /**
+   * @constructor
+   * @param {Object} options - Options for the listener
+   * @param {Array<multiaddr>} options.addresses
+   * @param {Array<Transport>} options.transports
+   * @param {Array<Multiplexer>} options.muxers
+   * @param {Array<Encryption>} options.encryption
+   * @param {PeerInfo} options.peerInfo
+   */
+  constructor ({
+    addresses = [multiaddr('/ip6/::/tcp/5892/ws')],
+    transports = [Transport],
+    muxers = [Muxer],
+    encryption = [Secio],
+    peerInfo
+  } = {}) {
+    this.peerAddr = addresses
     this.network = {}
     this.networkArray = []
-    this._emptyCachedDiscovery = this._cachedDiscovery = Buffer.from('01', 'hex')
+    this._cachedDiscovery = {}
 
-    this.switch.protos.addHandler('/p2p/stardust/0.1.0', (protocol, conn) => {
-      log('adding handler for stardust v0.1.0')
-
-      try {
-        this.handlerV010(conn, conn.info.info.muxed)
-      } catch (err) {
-        log(err)
-      }
-    })
+    this.libp2p = undefined
+    this._transports = transports
+    this._muxers = muxers
+    this._encryption = encryption
+    this._peerInfo = peerInfo
   }
 
-  async handler (conn) {
-    log('new connection')
+  removeFromNetwork (client) {
+    log('removing %s from network', client.id.toB58String())
+
+    delete this.network[client.id.toB58String()]
+    this.update()
+    this.broadcastDiscovery()
+  }
 
-    try {
-      const muxed = await this.switch.wrapInMuxer(conn, true) // add muxer ontop of raw socket
-      conn = await newMuxConn(muxed) // get a muxed connection
+  addToNetwork (connection, wrappedStream, id) {
+    log('adding %s to network', id.toB58String())
 
-      conn.muxed = muxed
-      await this.switch.negotiateProtocol(conn)
-    } catch (err) {
-      log(err)
+    const client = this.network[id.toB58String()] = {
+      id,
+      connection,
+      wrappedStream
     }
+
+    checkAckLoop(wrappedStream, () => this.removeFromNetwork(client))
+
+    this.update()
+    this.broadcastDiscovery()
   }
 
-  async handlerV010 (conn, muxed) {
-    const rpc = LP(conn) // turn into length-prefixed rpc interface
+  /**
+   * Update discovery data
+   */
+  update () {
+    log('updating cached data')
+    this.networkArray = Object.keys(this.network).map(b58 => this.network[b58])
+    this._cachedDiscovery = this.networkArray.length ? { ids: this.networkArray.map(client => client.id._id) } : { ids: [] }
+  }
 
-    try {
-      log('performing challenge')
+  /**
+   * Broadcast discovery data
+   */
+  broadcastDiscovery () {
+    log('broadcasting discovery to %o client(s)', this.networkArray.length)
+    this.networkArray.forEach(client => {
+      client.wrappedStream.writePB(this._cachedDiscovery, Discovery)
+    })
```

Comment:
yeah 👍 , I will create one issue

Reply:
The current behavior leans on that of the wrtc discovery module, which is why that is.

Also I optimized things so it does only one serialization of the entire list, but this means that we can't send different lists to different peers.

But I don't think that scale will be reached, so just sending different lists to different peers with just the newly joined ones should do it.

(There's a catch: If the peer count is zero a zero-length list (0x01) needs to be sent, otherwise the data that gets sent is zero-length - That is currently implemented)

Reply:
Created #19 to further discussion and track this improvement